### ISSUE TITLE
split machine values for kernel and userland

### ISSUE THREAD

**Issue by dcbaker** [mesonbuild/meson#4852]
In some cases the userland and the kernel are tightly coupled, like in windows. For other kernel's there are several different userlands in common use, examples include gnu/linux vs android/linux and macos/darwin vs ios/darwin. I'd like to propose adding at least one new api to the `machine` class, possibly two.

My for sure proposal is to add a `.userland()` method, that could be used as such:
```meson
if host_machine.userland() == 'gnu':
  ...
elif host_machine.userland() == 'android':
  ...
endif
```

I'd also like to propose adding a new api for the kernel `.kernel()` or `.os()` to replace `.system()`, but I'd be okay with just leaving `.system()` as well.

A related question is whether we want to consider different libc implementations as well. Some of them (musl notably) makes it very had to determine that you're using it in an automated fashion, so that might be a more difficult system to detect.

**Comment by nirbheek** [comment]
For userland autodetection, fwict either the toolchain knows about the userland (android/ios) in which case detection is reliable, or it doesn't (musl/dietlibc/uclibc). I think 90% of the use-cases would be satisfied with auto-detection for just Android and iOS, and the rest can specify it by hand in the native file or the cross file. We can always add more auto-detection as people ask.

Another thing I'm wondering about, is whether we should consider sub-classes of userland. For instance, Android has several different API levels with very different APIs and supported architectures, and watchOS and tvOS are distinct derivatives of iOS. We may want `userland` to mean the specific sub-type and `userland_family` to mean the generic type?

If we don't want to add sub-classes of userland, we can skip that and only implement `userland_family` for now.

**Comment by dcbaker** [comment]
That's a good point, I hadn't really thought about ios vs watchos or android (phone) vs androidtv, etc.

**Comment by dcbaker** [comment]
Closed in favor of #6361
